sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffleCoalesceExec.scala

```diff
@@ -279,22 +311,29 @@ class KudoTableOperator(kudo: Option[KudoSerializer], readOption: CoalesceReadOp
 abstract class HostCoalesceIteratorBase[T <: AutoCloseable : ClassTag](
     iter: Iterator[ColumnarBatch],
     targetBatchByteSize: Long,
-    metricsMap: Map[String, GpuMetric])
-  extends Iterator[CoalescedHostResult] with AutoCloseable {
-
-  private[this] val concatTimeMetric = metricsMap(GpuMetric.CONCAT_TIME)
-  private[this] val inputBatchesMetric = metricsMap(GpuMetric.NUM_INPUT_BATCHES)
-  private[this] val inputRowsMetric = metricsMap(GpuMetric.NUM_INPUT_ROWS)
+    concatTimeMetric: GpuMetric,
+    inputBatchesMetric: GpuMetric,
+    inputRowsMetric: GpuMetric,
+    asyncBuffering: Boolean = false
+) extends Iterator[CoalescedHostResult] with AutoCloseable {
   private[this] val serializedTables = new util.ArrayDeque[T]
-  private[this] var numTablesInBatch: Int = 0
-  private[this] var numRowsInBatch: Int = 0
-  private[this] var batchByteSize: Long = 0L
+  @volatile private[this] var numTablesInBatch: Int = 0
+  @volatile private[this] var numRowsInBatch: Int = 0
+  @volatile private[this] var batchByteSize: Long = 0L
+
+  private var executor: Option[ThrottlingExecutor] = None
 
   // Don't install the callback if in a unit test
   Option(TaskContext.get()).foreach { tc =>
     onTaskCompletion(tc)(close())
   }
 
+  // Don't try to call TaskContext.get().taskAttemptId() in the backend thread
+  private val taskAttemptID = Option(TaskContext.get()).
+    map(_.taskAttemptId().toString).getOrElse("unknown")
+
+  private var bufferingFuture : Option[Future[_]] = None
```

Comment:
```suggestion
  private var bufferingFuture: Option[Future[CoalescedHostResult]] = None
```

Reply:
we can't do this because it's not always a ColaescedHostResut, e.g. https://github.com/binmahone/spark-rapids/blob/5a1405e23a8099ad5db1a8197f61ce9b8e58a1b8/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffleCoalesceExec.scala#L376